### ISSUE TITLE
MissingSwitchDefault applied to switch expressions that return nothing

### ISSUE THREAD

**Issue by SiemelNaran** [checkstyle/checkstyle#10794]
> I have read check documentation: https://checkstyle.org/config_xxxxxx.html#NameOfAffectedCheck

The above link as https://checkstyle.org/config_xxxxxx.html#MissingSwitchDefault is not found on my browser.

> I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run

Yes.

> I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
<> [~/Downloads] [21:39:31] >> cat Sample.java 
/**
 * Useless Javadoc.
 */
public class Sample {
  private enum Options {
    ONE,
    TWO,
    THREE
  }

  private int error(Options option) {
    switch (option) { // line 12: checkstyle violation
      case ONE: return 1;
      case TWO: return 2;
      case THREE: return 3;
    }

    return 0;
  }

  private int ok1(Options option) {
    return switch (option) { // ok
      case ONE -> 1;
      case TWO -> 2;
      case THREE -> 3;
    };
  }

  private int ok2(Options option) {
    switch (option) { // line 30: should be ok
      case ONE -> System.out.println("1");
      case TWO -> System.out.println("2");
      case THREE -> System.out.println("3");
    };
    return 0;
  }
}
```

```bash
<> [~/Downloads] [21:42:53] >> java -version
openjdk version "16.0.2" 2021-07-20
OpenJDK Runtime Environment Temurin-16.0.2+7 (build 16.0.2+7)
OpenJDK 64-Bit Server VM Temurin-16.0.2+7 (build 16.0.2+7, mixed mode, sharing)

<> [~/Downloads] [21:43:05] >> javac -verbose Sample.java 2>&1 | grep -v loading
[parsing started SimpleFileObject[/Users/snaran/Downloads/Sample.java]]
[parsing completed 16ms]
[search path for source files: .]
[search path for class files: /Library/Java/JavaVirtualMachines/temurin-16.jdk/Contents/Home/lib/modules,.]
[checking Sample]
[wrote Sample$Options.class]
[wrote Sample$1.class]
[wrote Sample.class]
[total 234ms]

<> [~/Downloads] [21:43:10] >> java -jar checkstyle-9.0-all.jar -c /google_checks.xml Sample.java
Starting audit...
[WARN] /Users/snaran/Downloads/Sample.java:12:5: switch without "default" clause. [MissingSwitchDefault]
[WARN] /Users/snaran/Downloads/Sample.java:30:5: switch without "default" clause. [MissingSwitchDefault]
Audit done.
```

---

If a switch is a switch expression, then MissingSwitchDefault should not be applied, even if the switch expression does not return anything, as the compiler enforces that the switch is exhaustive.  It looks like checkstyle does not enforce the check for a switch expression that returns something (see ok1), but enforces check for a switch expression that does something.

**Comment by nmancus1** [comment]
From https://docs.oracle.com/javase/specs/jls/se16/html/jls-15.html#jls-15.28:

> ... expressions must produce a value for the result of the switch expression

Switch expressions must return a value, and be exhaustive as you've mentioned above.  However,  the switches in `error` and `ok2` are not switch expressions (they are switch statements), and therefore do not have to be exhaustive. This means that `MissingSwitchDefaultCheck` still applies; if a particular coding style guide requires code like this to have a default branch:

```java

    private int m(Options option) {
        int result = 0;
        switch (option) { // switch statement
            case ONE: result = 2;
            case TWO: result = 3;
            case THREE: result = 4;
        }
        return result;
    }

```

Then the rule should also apply to this, because they are semantically identical:

```java

    private int m(Options option) {
        int result = 0;
        switch (option) { // switch statement
            case ONE -> result = 2;
            case TWO -> result = 3;
            case THREE -> result = 4;
        }
        return result;
    }

```

And it does:

```bash

 ➜  src cat Test.java                                           
public class Test {
    private enum Options {
      ONE,
      TWO,
      THREE
    }

    private int m1(Options option) {
        int result = 0;
        switch (option) {
            case ONE -> result = 2;
            case TWO -> result = 3;
            case THREE -> result = 4;
        }
        return result;
    }

    private int m2(Options option) {
        int result = 0;
        switch (option) {
            case ONE -> result = 2;
            case TWO -> result = 3;
            case THREE -> result = 4;
        }
        return result;
    }
}
➜  src cat config.xml                                          
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name = "Checker">
        <module name="TreeWalker">
                <module name="MissingSwitchDefault">
                </module>
        </module>
</module>
 ➜  src java -jar checkstyle-9.0-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /home/nick/Desktop/Tester/src/Test.java:10:9: switch without "default" clause. [MissingSwitchDefault]
[ERROR] /home/nick/Desktop/Tester/src/Test.java:20:9: switch without "default" clause. [MissingSwitchDefault]
Audit done.
Checkstyle ends with 2 errors.

```

A few more examples to consider:

```java

 ➜  src cat Test.java                                           
public class Test {
    private enum Options {
      ONE,
      TWO,
      THREE
    }

    private int method1(Options option) {
      switch (option) { // switch statement, does not have to be exhaustive
        case ONE -> System.out.println("1");
        case TWO -> System.out.println("2");
        case THREE -> System.out.println("3");
        default -> System.out.println("default");
      }
      return 0;
    }

    private int method2(Options option) {
      switch (option) { // switch statement, does not have to be exhaustive, violation
        case ONE -> System.out.println("1");
        case TWO -> System.out.println("2");
      }
      return 0;
    }

    private int method3(Options option) {
        return switch (option) { // switch expression, must be exhaustive
            case ONE -> 2;
            case TWO -> 3;
            case THREE -> 4;
        };
    }

    private int method4(Options option) {
        final int result = switch (option) { // switch expression, must be exhaustive
            case ONE -> 2;
            case TWO -> 3;
            case THREE -> 4;
        };
        return result;
    }
}
 ➜  src cat config.xml                                          
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name = "Checker">
        <module name="TreeWalker">
                <module name="MissingSwitchDefault">
                </module>
        </module>
</module>
[8:44:20] ➜  src java -jar checkstyle-9.0-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /home/nick/Desktop/Tester/src/Test.java:19:7: switch without "default" clause. [MissingSwitchDefault]
Audit done.
Checkstyle ends with 1 errors.

```

**Comment by SiemelNaran** [comment]
Thx for explanation.  I didn't realize that switch expressions had to return something,  That document you posted says "It is a compile-time error if a switch expression has no result expressions" but strangely I got no compiler errors in the original code I posted.  So am closing out the issue.
